File: src-python/core/domain/ops_entry.py
"""操作日志领域对象。参见 PRD §2.6.4、DECISIONS D-0010。"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class OpsEntry:
    """单条操作日志。

    ops.jsonl 是业务关键依赖（D-0010），用于 undo 快照恢复、dirty 基线、fact 状态回放。
    op_id 是 Phase 1 必须字段。
    """

    op_id: str                          # 操作唯一 ID
    op_type: str                        # 操作类型（如 confirm_chapter / undo_chapter / edit_fact 等）
    target_id: str                      # 操作目标 ID（如 au_id / chapter_id / fact_id）
    timestamp: str                      # ISO 8601
    chapter_num: Optional[int] = None   # 关联章节号（可选）
    payload: dict[str, Any] = field(default_factory=dict)  # 操作负载（快照数据等）

File: src-engine/domain/ops_entry.ts
// Copyright (c) 2026 FicForge Contributors
// Licensed under the GNU Affero General Public License v3.0.

/** 操作日志领域对象。参见 PRD §2.6.4、DECISIONS D-0010。 */

export interface OpsEntry {
  op_id: string;                          // 操作唯一 ID
  op_type: string;                        // 操作类型
  target_id: string;                      // 操作目标 ID
  timestamp: string;                      // ISO 8601
  chapter_num: number | null;             // 关联章节号（可选）
  payload: Record<string, unknown>;       // 操作负载（快照数据等）
  device_id: string;                      // 设备唯一标识（同步用）
  lamport_clock: number;                  // 逻辑时钟（同步排序用）
}

export function createOpsEntry(partial: Pick<OpsEntry, "op_id" | "op_type" | "target_id" | "timestamp"> & Partial<OpsEntry>): OpsEntry {
  return {
    chapter_num: null,
    payload: {},
    device_id: "",
    lamport_clock: 0,
    ...partial,
  };
}
